### ISSUE TITLE
Aggregate public keys and signatures without jumping the C/Java boundary so often

### ISSUE THREAD

**Issue by benjaminion** [Consensys-Incorporated/teku#5182]
### Summary

We could probably significant reduce the cost of signature verification by adding some native code to aggregate public keys.

### Explanation

Currently our code for aggregating public keys looks like [this](https://github.com/ConsenSys/teku/blob/master/infrastructure/bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstPublicKey.java#L70):

```
    for (BlstPublicKey publicKey : publicKeys) {
      sum.add(publicKey.ecPoint);
    }
```

Each iteration of this is a separate JNI call to the Blst library, the cost of which probably dominates the cost of the relatively lightweight `add()` method.

We should
1. Profile this to see if it really is an issue. My gut feel is that there is a good speedup available here.
2. If so, add an external C library wrapper for Blst or fork of Blst that provides a method to aggregate a list of public keys, thus avoiding the calling overhead.

There is a [similar issue](https://github.com/ConsenSys/teku/blob/master/infrastructure/bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java#L61) for signatures, but for a number of reasons it's probably not so significant: we don't aggregate signatures as often; the cost of addition (aggregation) of signatures is greater; and the signature aggregation does an expensive group check, unlike the public key aggregation. So the JNI overhead would be relatively lower. But if we're adding native code for public keys, we could also do it for signatures for completeness.

**Comment by benjaminion** [comment]
A quick and dirty update on testing this. It turns out that SWIG is a monster, so coding up the full version of this is complex. For a simple feasibility study I made two versions of point addition:

```
    P1* add(const P1_Affine &a)
    {   blst_p1_add_or_double_affine(&point, &point, a); return this;   }
    P1* add2(const P1_Affine &a, const P1_Affine &b)
    {   blst_p1_add_or_double_affine(&point, &point, a); blst_p1_add_or_double_affine(&point, &point, b); return this;   }
```

Given an even number of points the second will have half as many JNI calls as the first. Running with ten million points I get the following timings:
  - `add`: 7.5s
  - `add2`: 7.2 s

Thus, 5m JNI calls takes 0.3s, for a per-call cost of 0.06 microseconds, and 10m adds without the calling overhead takes 6.9s, for a per-add cost of 0.69 microseconds.

In summary, we could likely get around an 8% speed up in public key aggregation by eliminating the calling overhead. This is borderline worth doing. A worst-case verification of a block with 64 full attestations would need around 20,000 of these point additions which would currently take around 0.015s. We could shave just over a millisecond off this by eliminating the calling overhead.

The method that adds multiple points at once is already available in Blst and is exposed via some language interfaces, just not Java. In principle it's easy to expose it, except that SWIG is giving me a headache, and we'd have to maintain a branch of Blst or push the change back upstream.

Leaving this for now, but may come back to it some time.

**Comment by tbenr** [comment]
I did some additional tests implementing various version of `add`, including a dynamic `add` accepting an array.
The dynamic one seems not paying off due to additional array handling.

I focused then on:

```
add2loop: 0.0019 ms
add2: 0.0017 ms - gain: 11.764707%

add4loop: 0.0036 ms
add4: 0.0031 ms - gain: 16.129026%

add8loop: 0.0071 ms
add8: 0.006 ms - gain: 18.333334%

add16loop: 0.0142 ms
add16: 0.0118 ms - gain: 20.33898%

add32loop: 0.0282 ms
add32: 0.0233 ms - gain: 21.030048%

add64loop: 0.0563 ms
add64: 0.0464 ms - gain: 21.336206%

add128loop: 0.1123 ms
add128: 0.0922 ms - gain: 21.80043%

** block simulation (360 * 69)
add: 22.0 ms
agg add: 18.0 ms - gain: 22.222221%
```

Seem the benefits reaches the top on `add32`. No additional benefits with more parameters.

At time of writing, after profiling a node for 10 min:
![image](https://github.com/Consensys/teku/assets/15999009/3ccd825d-97ed-4e92-aade-dc876ca29661)
We can gain 20% on the `P1.add`.

So I think the added complexity seems not paid back by the performance boost.


Giving up.

**Comment by benjaminion** [comment]
Yep, the gains don't seem worth it. Or, to look at it more positively, JNI overhead is not currently a significant issue for us. Thanks for looking into this @tbenr - closing now.
